aii-opennebula/src/main/pan/quattor/aii/opennebula/functions.pan

```diff
@@ -0,0 +1,81 @@
+declaration template quattor/aii/opennebula/functions;
+
+include 'pan/types';
+
+@documentation{
+This function generates OpenNebula MAC addresses from MAC_PREFIX + IPv4
+Based on OpenNebula opennebula_ipv42mac function:
+
+https://github.com/OpenNebula/one/blob/master/share/router/vmcontext.rb
+
+Syntax: 
+mac_prefix:string ipv4:string
+
+mac_prefix hex:hex value used also by oned.conf (02:00 by default)
+ipv4 IP used by the VM
+}
+function opennebula_ipv42mac = {
+    # Check arguments
+    if (ARGC != 2) {
+        error("usage: \"hwaddr\" = opennebula_ipv42mac(MAC_PREFIX, IP)");
+    };
+    # Sanity check
+    if (!match(ARGV[0],
+        '^[0-9a-f]{2}[:][0-9a-f]{2}$')) {
+        error(format("Invalid MAC_PREFIX format: %s", ARGV[0]));
+    };
+    if (!is_ipv4(ARGV[1])) {
+        error(format("Invalid IPv4 format: %s", ARGV[1]));
+    };
+    # Convert IP octets to Hex
+    ipoctets = list();
+    foreach (i; octet; split('\.',ARGV[1])) {
+        ipoctets[i] = format("%02x", to_long(octet));
+    };
+    return(format("%s:%s:%s:%s:%s", ARGV[0], ipoctets[0], ipoctets[1], ipoctets[2], ipoctets[3]));
```

Comment:
@alvarosimon Shouldn't be %s:%s:%s:%s:%s:%s ?
A mac address is composed by 6 pairs and not 5

Anyway, we managed to write our join function. IMHO, this kind of functions should be ported back to quattor because they could be useful.